File: backend/app/container.py
from app.application.use_cases.account.change_password import ChangePassword
from app.application.use_cases.account.delete_user import DeleteUser
from app.application.use_cases.analytics.retrieve_distribution import (
    RetrieveDistribution,
)
from app.application.use_cases.analytics.retrieve_time_series import RetrieveTimeSeries
from app.application.use_cases.auth.get_user_id_from_token import GetUserIdFromToken
from app.application.use_cases.auth.login import Login
from app.application.use_cases.auth.register_user import RegisterUser
from app.application.use_cases.speeches.delete_speech import DeleteSpeech
from app.application.use_cases.speeches.list_speeches import ListSpeeches
from app.application.use_cases.speeches.process_speech import ProcessSpeech
from app.infrastructure.composition import InfrastructureComposition
from app.settings import InfrastructureSettings


class ApplicationContainer:
    def __init__(self, settings: InfrastructureSettings) -> None:
        self.infrastructure = InfrastructureComposition(settings)
        self.change_password = ChangePassword(
            self.infrastructure.uow_factory,
            self.infrastructure.password_hasher,
        )
        self.delete_user = DeleteUser(self.infrastructure.uow_factory)
        self.retrieve_distribution = RetrieveDistribution(
            self.infrastructure.uow_factory
        )
        self.retrieve_time_series = RetrieveTimeSeries(self.infrastructure.uow_factory)
        self.get_user_id_from_token = GetUserIdFromToken(
            self.infrastructure.token_service, self.infrastructure.uow_factory
        )
        self.login = Login(
            self.infrastructure.uow_factory,
            self.infrastructure.password_hasher,
            self.infrastructure.token_service,
        )
        self.register_user = RegisterUser(
            self.infrastructure.uow_factory, self.infrastructure.password_hasher
        )
        self.delete_speech = DeleteSpeech(self.infrastructure.uow_factory)
        self.list_speeches = ListSpeeches(self.infrastructure.uow_factory)
        self.process_speech = ProcessSpeech(
            self.infrastructure.uow_factory,
            self.infrastructure.grammar_analyzer,
            self.infrastructure.analysis_quota_enforcer,
        )

    async def close(self) -> None:
        await self.infrastructure.close()

File: src/bootstrap/container.ts
import "server-only";

import { Redis } from "@upstash/redis";

import {
  CleanupAbandonedAudio,
  ChangePassword,
  DeleteSpeech,
  DeleteUser,
  ListSpeeches,
  Login,
  ProcessSpeech,
  RegisterAndStartSession,
  ResolveSession,
  RetrieveAnalyticsDashboard,
} from "@/src/application/use-cases";
import { Argon2PasswordHasher } from "@/src/infrastructure/auth/argon2-password-hasher";
import { JwtTokenService } from "@/src/infrastructure/auth/jwt-token-service";
import {
  VercelBlobStagedAudioJanitor,
  VercelBlobStagedAudioStore,
} from "@/src/infrastructure/blob";
import { getServerEnvironment } from "@/src/infrastructure/config/env";
import {
  DeterministicGrammarAnalyzer,
  DeterministicTranscriber,
} from "@/src/infrastructure/openai/deterministic";
import { OpenAiClient } from "@/src/infrastructure/openai/client";
import { OpenAiGrammarAnalyzer } from "@/src/infrastructure/openai/grammar-analyzer";
import { OpenAiTranscriber } from "@/src/infrastructure/openai/transcriber";
import {
  PostgresAnalyticsReader,
  PostgresSpeechRepository,
  PostgresUserRepository,
} from "@/src/infrastructure/postgres";
import { UpstashAnalysisQuota } from "@/src/infrastructure/quota";
import { getLogger } from "@/src/infrastructure/observability/logger";

export interface ApplicationContainer {
  readonly registerAndStartSession: RegisterAndStartSession;
  readonly login: Login;
  readonly resolveSession: ResolveSession;
  readonly changePassword: ChangePassword;
  readonly deleteUser: DeleteUser;
  readonly processSpeech: ProcessSpeech;
  readonly listSpeeches: ListSpeeches;
  readonly deleteSpeech: DeleteSpeech;
  readonly retrieveAnalyticsDashboard: RetrieveAnalyticsDashboard;
  readonly cleanupAbandonedAudio: CleanupAbandonedAudio;
}

let applicationContainer: ApplicationContainer | undefined;

/** The application's only composition root. No domain or use-case imports a framework. */
export function getApplicationContainer(): ApplicationContainer {
  applicationContainer ??= createApplicationContainer();
  return applicationContainer;
}

export function createApplicationContainer(): ApplicationContainer {
  const environment = getServerEnvironment();

  const users = new PostgresUserRepository();
  const speeches = new PostgresSpeechRepository();
  const analytics = new PostgresAnalyticsReader();
  const passwordHasher = new Argon2PasswordHasher();
  const tokens = new JwtTokenService({
    secret: environment.jwtSecret,
    expirationMinutes: environment.sessionTtlMinutes,
  });

  const quota = new UpstashAnalysisQuota(
    new Redis({
      url: environment.upstashRedisRestUrl,
      token: environment.upstashRedisRestToken,
    }),
    {
      minuteLimit: environment.analysisMinuteLimit,
      dayLimit: environment.analysisDayLimit,
    },
  );

  let transcriber;
  let grammarAnalyzer;
  if (environment.analyzerMode === "deterministic") {
    transcriber = new DeterministicTranscriber();
    grammarAnalyzer = new DeterministicGrammarAnalyzer();
  } else {
    const apiKey = environment.openAiApiKey;
    if (!apiKey) {
      // The environment parser owns this invariant. Keep the composition root
      // safe if a future configuration source bypasses that parser.
      throw new Error("OPENAI_API_KEY is required");
    }
    const client = new OpenAiClient({
      apiKey,
      baseUrl: environment.openAiBaseUrl,
      timeoutMs: environment.openAiTimeoutMs,
      maxRetries: environment.openAiMaxRetries,
    });
    transcriber = new OpenAiTranscriber(
      client,
      environment.openAiTranscriptionModel,
    );
    grammarAnalyzer = new OpenAiGrammarAnalyzer(
      client,
      environment.openAiAnalysisModel,
    );
  }

  const stagedAudioStore = new VercelBlobStagedAudioStore(
    environment.blobStoreId,
  );
  const stagedAudioJanitor = new VercelBlobStagedAudioJanitor(
    environment.blobStoreId,
  );

  return Object.freeze({
    registerAndStartSession: new RegisterAndStartSession(
      users,
      passwordHasher,
      tokens,
    ),
    login: new Login(users, passwordHasher, tokens),
    resolveSession: new ResolveSession(tokens, users),
    changePassword: new ChangePassword(users, passwordHasher),
    deleteUser: new DeleteUser(users),
    processSpeech: new ProcessSpeech(
      stagedAudioStore,
      speeches,
      transcriber,
      grammarAnalyzer,
      quota,
      (error) => {
        getLogger().error(
          {
            operation: "delete_staged_audio",
            errorType: error instanceof Error ? error.name : "UnknownError",
          },
          "Staged audio cleanup failed",
        );
      },
    ),
    listSpeeches: new ListSpeeches(speeches),
    deleteSpeech: new DeleteSpeech(speeches),
    retrieveAnalyticsDashboard: new RetrieveAnalyticsDashboard(analytics),
    cleanupAbandonedAudio: new CleanupAbandonedAudio(stagedAudioJanitor),
  });
}
